RNA_folding.py

```diff
@@ -0,0 +1,319 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from os.path import dirname, join
+from itertools import product, combinations
+from collections import defaultdict
+import numpy as np
+import networkx as nx
+import matplotlib.pyplot as plt
+import dimod
+import click
+from dwave.system import LeapHybridCQMSampler
+
+
+def text_to_matrix(file_name, min_loop=2):
+    """ Reads properly formatted RNA text file and returns a matrix of possible hydrogen bonding pairs.
+
+    Args:
+        file_name: String path to text file.
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+    """
+
+    # Requires text file of rna data written in same format as examples.
+    with open(file_name) as f:
+        rna = "".join(("".join(line.split()[1:]) for line in f.readlines())).lower()
+
+    # Create a dictionary of all indices where each nucleotide occurs.
+    index_dict = defaultdict(list)
+
+    # create a dictionary giving list of indices for each nucleotide.
+    for i, nucleotide in enumerate(rna.lower()):
+        index_dict[nucleotide].append(i)
+
+    # List of possible hydrogen bonds for stems.
+    # Recall that 't' is just a stand in for 'u'.
+    hydrogen_bonds = [('a', 't'), ('a', 'u'), ('c', 'g'), ('g', 't'), ('g', 'u')]
+
+    # Create a upper triangular 0/1 matrix indicated where bonds may occur.
+    bond_matrix = np.zeros((len(rna), len(rna)), dtype=bool)
+    for pair in hydrogen_bonds:
+        for bond in product(index_dict[pair[0]], index_dict[pair[1]]):
+            if abs(bond[0] - bond[1]) > min_loop:
+                bond_matrix[min(bond), max(bond)] = 1
+
+    return bond_matrix
+
+
+def make_stem_dict(bond_matrix, min_stem=3, min_loop=2):
+    """ Takes a matrix of potential hydrogen binding pairs and returns a dictionary of possible stems.
+
+    The stem dictionary records the maximal stems (under inclusion) as keys,
+    where each key maps to a list of the associated stems weakly contained within the maximal stem.
+    Recording stems in this manner allows for faster computations.
+
+    Args:
+        bond_matrix: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+        min_stem: Integer minimum number of nucleotides in each side of a stem.
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Dictionary of all possible stems with maximal stems as keys.
+    """
+
+    stem_dict = {}
+    n = bond_matrix.shape[0]
+
+    # Iterate through matrix looking for possible stems.
+    for i in range(n - (2 * min_stem + min_loop)):
```

Comment:
Should this have a `+1`?  For example, suppose I have `n=10` nuceotides, `min_stem=5`, and `min_loop=0`.  This will produce an empty `for` loop, but is the length-5 stem not valid?
```suggestion
    for i in range(n - (2 * min_stem + min_loop) + 1):
```